src/linear_algebra/matrix/is_diag.lean

```diff
@@ -0,0 +1,197 @@
+
+/-- `A.is_diag` means square matrix `A` is a dianogal matrix: `∀ i j, i ≠ j → A i j = 0`. -/
+def is_diag [has_zero α] (A : matrix n n α) : Prop := ∀ ⦃i j⦄, i ≠ j → A i j = 0
+
+/-- Matrix `A` is diagonal iff there is a vector `d`
+    such that A is the diagonal matix generated by `d`. -/
+lemma is_diag_iff_eq_diagonal [has_zero α] [decidable_eq n] (A : matrix n n α) :
+  A.is_diag ↔ (∃ d, A = diagonal d) :=
+begin
+  split,
+  { intros h,
+    use (λ i, A i i),
+    ext i j,
+    by_cases g : i = j;
+    simp [@h i j, g] },
+  { rintros ⟨d, rfl⟩ i j h, simp [h] }
+end
+
+/-- Every matrix indexed by a subsingleton is diagonal. -/
+lemma is_diag_of_subsingleton [has_zero α] [subsingleton n] (A : matrix n n α) : A.is_diag :=
+by { intros i j h, have h' := subsingleton.elim i j, contradiction }
+
+/-- Every zero matrix is diagonal. -/
+@[simp] lemma is_diag_zero [has_zero α] : (0 : matrix n n α).is_diag :=
+λ i j h, rfl
+
+/-- Every identity matrix is diagonal. -/
+@[simp] lemma is_diag_one [decidable_eq n] [has_zero α] [has_one α] :
+  (1 : matrix n n α).is_diag :=
+λ i j, one_apply_ne
+
+lemma is_diag.map [has_zero α] [has_zero β]
+{A : matrix n n α} (ha : A.is_diag) {f : α → β} (hf : f 0 = 0) :
+  (A.map f).is_diag :=
+by { intros i j h, simp [ha h, hf] }
+
+lemma is_diag.neg [add_group α] {A : matrix n n α} (ha : A.is_diag) :
+  (-A).is_diag :=
+by { intros i j h, simp [ha h] }
+
+@[simp] lemma is_diag_neg_iff [add_group α] {A : matrix n n α} :
+  (-A).is_diag ↔ A.is_diag :=
+⟨ λ ha i j h, neg_eq_zero.1 (ha h), is_diag.neg ⟩
+
+lemma is_diag.add
+  [add_zero_class α] {A B : matrix n n α} (ha : A.is_diag) (hb : B.is_diag) :
+  (A + B).is_diag :=
+by { intros i j h, simp [ha h, hb h] }
+
+lemma is_diag.sub [add_group α]
+  {A B : matrix n n α} (ha : A.is_diag) (hb : B.is_diag) :
+  (A - B).is_diag :=
+by { intros i j h, simp [ha h, hb h] }
+
+@[simp] lemma is_diag.smul [monoid R] [add_monoid α] [distrib_mul_action R α]
+  {k : R} {A : matrix n n α} (ha : A.is_diag) :
+  (k • A).is_diag :=
+by { intros i j h, simp [ha h] }
+
+lemma is_diag.transpose [has_zero α] {A : matrix n n α} (ha : A.is_diag) : Aᵀ.is_diag :=
+λ i j h, ha h.symm
+
+@[simp] lemma is_diag_transpose_iff [has_zero α] {A : matrix n n α} :
+  Aᵀ.is_diag ↔ A.is_diag :=
+⟨ is_diag.transpose, is_diag.transpose ⟩
+
+lemma is_diag.conj_transpose
+  [semiring α] [star_ring α] {A : matrix n n α} (ha : A.is_diag) :
+  Aᴴ.is_diag :=
+ha.transpose.map (star_zero _)
+
+@[simp] lemma is_diag_conj_transpose_iff [semiring α] [star_ring α] {A : matrix n n α} :
+  Aᴴ.is_diag ↔ A.is_diag :=
+⟨ λ ha, by {convert ha.conj_transpose, simp}, is_diag.conj_transpose ⟩
+
+lemma is_diag.minor [has_zero α]
+  {A : matrix n n α} (ha : A.is_diag) {f : m → n} (hf : injective f) :
+  (A.minor f f).is_diag :=
+λ i j h, ha (hf.ne h)
+
+/-- `(A ⊗ B).is_diag` if both `A` and `B` are diagonal. -/
+lemma is_diag.kronecker [mul_zero_class α]
+  {A : matrix m m α} {B : matrix n n α} (hA: A.is_diag) (hB: B.is_diag) :
```

Comment:
Spaces around `:`, again. Please can you try to be more careful about this in future PRs, it's tiring having to keep spotting these in a review. You can use a regex search of `(?<! ):(?! |=)` to find all of these in vs-code.
```suggestion
  {A : matrix m m α} {B : matrix n n α} (hA : A.is_diag) (hB : B.is_diag) :
```

Reply:
Sure